File: back/src/gamengine/schemas.go
package gamengine

import (
	"github.com/diyor28/not-agar/src/csbin"
	"github.com/diyor28/not-agar/src/gamengine/constants"
	_map "github.com/diyor28/not-agar/src/gamengine/map"
	"github.com/diyor28/not-agar/src/gamengine/map/entity"
	"github.com/diyor28/not-agar/src/gamengine/map/food"
	"github.com/diyor28/not-agar/src/gamengine/map/players"
	"github.com/diyor28/not-agar/src/gamengine/map/players/shell"
	"github.com/diyor28/not-agar/src/gamengine/map/spikes"
	"reflect"
)

var playerField = csbin.NewField("player", reflect.Struct).SubFields(
	csbin.NewField("x", reflect.Float32),
	csbin.NewField("y", reflect.Float32),
	csbin.NewField("weight", reflect.Float32),
	csbin.NewField("nickname", reflect.String).MaxLen(255),
	csbin.NewField("color", reflect.Array).Len(3).SubType(csbin.NewField("color", reflect.Uint8)),
)

var spikeField = csbin.NewField("spike", reflect.Struct).SubFields(
	csbin.NewField("x", reflect.Float32),
	csbin.NewField("y", reflect.Float32),
	csbin.NewField("weight", reflect.Float32),
)

var foodField = csbin.NewField("food", reflect.Struct).SubFields(
	csbin.NewField("id", reflect.Uint32),
	csbin.NewField("x", reflect.Float32),
	csbin.NewField("y", reflect.Float32),
	csbin.NewField("weight", reflect.Float32),
	csbin.NewField("color", reflect.Array).Len(3).SubType(csbin.NewField("color", reflect.Uint8)),
)

var GenericSchema = csbin.New(
	csbin.NewField("event", reflect.Uint8),
)

type GenericEvent struct {
	Event constants.GameEvent
}

var PingPongSchema = GenericSchema.Extends(
	csbin.NewField("timestamp", reflect.Uint64),
)

var StartSchema = GenericSchema.Extends(
	csbin.NewField("nickname", reflect.String).MaxLen(255),
)

var StartedSchema = GenericSchema.Extends(
	csbin.NewField("player", reflect.Struct).SubFields(
		csbin.NewField("x", reflect.Float32),
		csbin.NewField("y", reflect.Float32),
		csbin.NewField("weight", reflect.Float32),
		csbin.NewField("color", reflect.Array).Len(3).SubType(csbin.NewField("color", reflect.Uint8)),
		csbin.NewField("points", reflect.Slice).MaxLen(255).SubType(csbin.NewField("point", reflect.Struct).SubFields(
			csbin.NewField("x", reflect.Float32),
			csbin.NewField("y", reflect.Float32),
		)),
	),
	csbin.NewField("spikes", reflect.Slice).MaxLen(255).SubType(spikeField),
	csbin.NewField("food", reflect.Slice).MaxLen(10000).SubType(foodField),
)

type StartedEventPlayer struct {
	X      float32
	Y      float32
	Weight float32
	Color  [3]uint8
	Points []*shell.Point
}

type StartedEvent struct {
	Event  constants.GameEvent
	Player *StartedEventPlayer
	Spikes []*spikes.Spike
	Food   []*food.Food
}

var MoveSchema = GenericSchema.Extends(
	csbin.NewField("newX", reflect.Float32),
	csbin.NewField("newY", reflect.Float32),
)

type MoveEvent struct {
	Event constants.GameEvent
	NewX  float32 `json:"newX"`
	NewY  float32 `json:"newY"`
}

var MovedSchema = GenericSchema.Extends(
	csbin.NewField("x", reflect.Float32),
	csbin.NewField("y", reflect.Float32),
	csbin.NewField("weight", reflect.Float32),
	csbin.NewField("velocityX", reflect.Float32),
	csbin.NewField("velocityY", reflect.Float32),
	csbin.NewField("zoom", reflect.Float32),
	csbin.NewField("points", reflect.Slice).MaxLen(255).SubType(
		csbin.NewField("point", reflect.Struct).SubFields(
			csbin.NewField("x", reflect.Float32),
			csbin.NewField("y", reflect.Float32),
		)),
)

type MovedEvent struct {
	Event     constants.GameEvent
	X         float32
	Y         float32
	VelocityX float32
	VelocityY float32
	Weight    float32
	Zoom      float32
	Points    []*shell.Point
}

var PlayerStatsSchema = GenericSchema.Extends(
	csbin.NewField("topPlayers", reflect.Slice).MaxLen(255).SubType(csbin.NewField("players", reflect.Struct).SubFields(
		csbin.NewField("nickname", reflect.String).MaxLen(255),
		csbin.NewField("weight", reflect.Int16),
	)),
)

type PlayerStatsEvent struct {
	Event      constants.GameEvent
	TopPlayers []_map.PlayerStat
}

var AdminStatsSchema = GenericSchema.Extends(
	csbin.NewField("botsCount", reflect.Uint16),
	csbin.NewField("playersCount", reflect.Uint16),
	csbin.NewField("topsPlayers", reflect.Slice).MaxLen(255).SubType(playerField),
)

var FoodCreatedSchema = GenericSchema.Extends(
	csbin.NewField("food", reflect.Slice).MaxLen(10000).SubType(foodField),
)

type FoodCreatedEvent struct {
	Event constants.GameEvent
	Food  []*food.Food
}

var FoodEatenSchema = GenericSchema.Extends(
	csbin.NewField("id", reflect.Uint32),
)

type FoodEatenEvent struct {
	Event constants.GameEvent
	Id    entity.Id
}

var PlayersUpdatedSchema = GenericSchema.Extends(
	csbin.NewField("players", reflect.Slice).MaxLen(255).SubType(playerField),
)

type PlayersUpdatedEvent struct {
	Event   constants.GameEvent
	Players []*players.Player
}

File: front/src/client/schemas.ts
import {Schema} from "../codec";

export enum GameEvent {
	Ping,
	Pong,
	Move,
	Moved,
	Start,
	Started,
	FoodEaten,
	FoodCreated,
	PlayersUpdate,
	StatsUpdate,
	Rip
}

export const genericSchema = new Schema({
	event: 'uint8'
});

export const pingSchema = genericSchema.extends({
	timestamp: 'uint64'
});

export const moveSchema = genericSchema.extends({
	newX: 'float32',
	newY: 'float32'
});

export const movedSchema = genericSchema.extends({
	x: 'float32',
	y: 'float32',
	weight: 'float32',
	velocityX: 'float32',
	velocityY: 'float32',
	zoom: 'float32',
	points: {
		type: 'array',
		of: {x: 'int16', y: 'int16'},
		maxLen: 255
	}
});

export const startSchema = genericSchema.extends({
	nickname: {type: 'string', maxLen: 255}
});

export const startedSchema = genericSchema.extends({
	player: {
		x: 'float32',
		y: 'float32',
		weight: 'float32',
		color: {type: 'array', of: 'uint8', length: 3},
		points: {
			type: 'array',
			of: {x: 'int16', y: 'int16'},
			maxLen: 255
		}
	},
	spikes: {
		type: 'array',
		of: {
			x: 'float32',
			y: 'float32',
			weight: 'float32'
		},
		maxLen: 255
	},
	food: {
		type: 'array',
		of: {
			id: 'uint32',
			x: 'float32',
			y: 'float32',
			weight: 'float32',
			color: {type: 'array', of: 'uint8', length: 3}
		},
		maxLen: 10_000
	}
});

export const foodCreatedSchema = genericSchema.extends({
	food: {
		type: 'array',
		of: {
			id: 'uint32',
			x: 'float32',
			y: 'float32',
			weight: 'float32',
			color: {type: 'array', of: 'uint8', length: 3}
		},
		maxLen: 10_000
	}
});

export const foodEatenSchema = genericSchema.extends({
	id: 'uint32'
});

export const playersSchema = genericSchema.extends({
	players: {
		type: 'array',
		of: {
			x: 'uint16',
			y: 'uint16',
			weight: 'float32',
			nickname: {type: 'string', maxLen: 255},
			color: {type: 'array', of: 'uint8', length: 3}
		},
		maxLen: 255
	}
});

export const statsSchema = genericSchema.extends({
	topPlayers: {
		type: 'array',
		of: {
			nickname: {type: 'string', maxLen: 255},
			weight: 'int16'
		},
		maxLen: 255
	}
});

export const ripSchema = genericSchema;
